From 7a345f9d8927d9596849ed5c0511f51be7062bd3 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Wed, 3 Apr 2019 13:01:52 -0700
Subject: [PATCH 0604/1852] v4l: xilinx: scd: Request IRQ after completing
 initialization

commit 915e0b67cf9bea8ada586c754b10243004225e93 from
https://github.com/Xilinx/linux-xlnx.git

IRQs can be triggered as soon as they're requested. To ensure a valid
state in the IRQ handler, make sure to request the IRQ after
initializing all components.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../platform/xilinx/xilinx-scenechange.c      | 20 ++++++++++---------
 .../platform/xilinx/xilinx-scenechange.h      |  1 -
 2 files changed, 11 insertions(+), 10 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.c b/drivers/media/platform/xilinx/xilinx-scenechange.c
index 9d9d400cb889..8feada6703d1 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.c
@@ -43,6 +43,12 @@ static int xscd_init_resources(struct xscd_device *xscd)
 	if (IS_ERR(xscd->iomem))
 		return PTR_ERR(xscd->iomem);
 
+	xscd->irq = platform_get_irq(pdev, 0);
+	if (xscd->irq < 0) {
+		dev_err(xscd->dev, "No valid irq found\n");
+		return -EINVAL;
+	}
+
 	xscd->clk = devm_clk_get(xscd->dev, NULL);
 	if (IS_ERR(xscd->clk))
 		return PTR_ERR(xscd->clk);
@@ -76,15 +82,6 @@ static int xscd_parse_of(struct xscd_device *xscd)
 		return -EINVAL;
 	}
 
-	xscd->irq = irq_of_parse_and_map(node, 0);
-	if (!xscd->irq) {
-		dev_err(xscd->dev, "No valid irq found\n");
-		return -EINVAL;
-	}
-
-	ret = devm_request_irq(xscd->dev, xscd->irq, xscd_irq_handler,
-			       IRQF_SHARED, dev_name(xscd->dev), xscd);
-
 	return 0;
 }
 
@@ -138,6 +135,11 @@ static int xscd_probe(struct platform_device *pdev)
 	if (ret < 0)
 		dev_err(&pdev->dev, "Failed to initialize the DMA\n");
 
+	ret = devm_request_irq(xscd->dev, xscd->irq, xscd_irq_handler,
+			       IRQF_SHARED, dev_name(xscd->dev), xscd);
+	if (ret < 0)
+		dev_err(&pdev->dev, "Failed to request IRQ\n");
+
 	dev_info(xscd->dev, "scene change detect device found!\n");
 	return 0;
 }
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.h b/drivers/media/platform/xilinx/xilinx-scenechange.h
index d18084804719..18b4d8a8b886 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.h
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.h
@@ -19,7 +19,6 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/of.h>
-#include <linux/of_irq.h>
 #include <linux/platform_device.h>
 #include <linux/xilinx-v4l2-controls.h>
 
-- 
2.31.1

